storage-xray-udp/src/main/java/zipkin2/storage/xray_udp/UDPMessageEncoder.java

```diff
@@ -42,9 +42,17 @@ static void writeJson(Span span, Buffer buffer) throws IOException {
       writer.name("type").value("subsegment");
       if (span.kind() != null) writer.name("namespace").value("remote");
       // some remote service's name can be null, using null names causes invisible subsegment
-      // using "unknown" subsegment name will help to detect missing names
-      writer.name("name")
-          .value(span.remoteServiceName() == null ? "unknown" : span.remoteServiceName());
+      // using "unknown" subsegment name will help to detect missing names but will also
+      // result in the service map displaying services depending on an "unknown" one.
+      if (span.remoteServiceName() == null) {
+        if ("true".equalsIgnoreCase(System.getenv("AWS_XRAY_USE_LOCAL_SPAN_FOR_MISSING_REMOTES"))) {
```

Comment:
hmm this would be the only time we use environment variables directly in java code of a component. Let's not do that. At worst a flag, but this also seems bad semantically. Also, localServiceName can be null in odd situations (not in brave, but possible if people are writing spans themselves).

@devinsba since you use this code what do you think? Is there another option? What's the impact of the graph not having a correct remote service name?

I suppose it would be telling to see a trace where this happens, too.. When thinking about tests, I was actually more thinking that but didn't mention it. Sometimes seeing the trace data can show what options might not be obvious from looking at the code without any context.

Reply:
I am not a big fan about using environment variables in java code, but as the address of the X-Ray server is passed as `AWS_XRAY_DAEMON_ADDRESS` env var I used the same for consistency.

Having `unknown` as X-Ray subsegment name creates the issues described here: https://github.com/openzipkin/zipkin-aws/issues/87. I believe that unknown segments when looking at a single trace are not a big issue, but having almost all services reported as depending on unknown in the service map is a problem if the main reason for using Zipkin or X-Ray is to have a map of the service dependencies. 

I have copy/pasted examples of a trace (received from a Spring Boot application that uses the Spring Cloud stack, OpenFeign and Hystrix) and the related data sent to X-Ray, will do it again here:

## First span (the request is received)
```
{
  "traceId": "5b07b67fed3a0b4570cbdfe07b779e32",
  "id": "70cbdfe07b779e32",
  "kind": "SERVER",
  "name": "get",
  "timestamp": 1527232127970111,
  "duration": 29131,
  "localEndpoint": {
    "serviceName": "security-gateway-aws",
    "ipv4": "192.168.0.171"
  },
  "remoteEndpoint": {
    "ipv4": "127.0.0.1",
    "port": 33060
  },
  "tags": {
    "http.method": "GET",
    "http.path": "/test-app/v1/public"
  }
}
```
and the related X-Ray segment:
```
{
  "trace_id": "1-5b07b67f-ed3a0b4570cbdfe07b779e32",
  "id": "70cbdfe07b779e32",
  "name": "security-gateway-aws",
  "start_time": 1.527232127970111E9,
  "end_time": 1.527232127999242E9,
  "http": {
    "request": {
      "method": "GET"
    }
  },
  "annotations": {
    "operation": "get",
    "http_path": "/test-app/v1/public"
  }
}
```
No problems here.

## Second span (the service invokes a remote API via OpenFeign)
```
{
  "traceId": "5b07b67fed3a0b4570cbdfe07b779e32",
  "parentId": "70cbdfe07b779e32",
  "id": "02e913691d305aa0",
  "kind": "CLIENT",
  "name": "get",
  "timestamp": 1527232127972240,
  "duration": 23838,
  "localEndpoint": {
    "serviceName": "security-gateway-aws",
    "ipv4": "192.168.0.171"
  },
  "tags": {
    "http.method": "GET",
    "http.path": "/test-app/v1/public"
  }
}
```
and the related X-Ray subsegment:
```
{
  "trace_id": "1-5b07b67f-ed3a0b4570cbdfe07b779e32",
  "parent_id": "70cbdfe07b779e32",
  "id": "02e913691d305aa0",
  "type": "subsegment",
  "namespace": "remote",
  "name": "unknown",
  "start_time": 1.52723212797224E9,
  "end_time": 1.527232127996078E9,
  "http": {
    "request": {
      "method": "GET"
    }
  },
  "annotations": {
    "operation": "get",
    "http_path": "/test-app/v1/public"
  }
}
```
Besides the conceptual issue that this is not a remote call, yet (but Zipkin cannot know this) here the first time the unknown problem appears.

## Third span (Hystrix circuit breaker used by OpenFeign)
```
{
  "traceId": "5b07b67fed3a0b4570cbdfe07b779e32",
  "parentId": "02e913691d305aa0",
  "id": "f3d698d7b353651f",
  "name": "hystrix",
  "timestamp": 1527232127972476,
  "duration": 23595,
  "localEndpoint": {
    "serviceName": "security-gateway-aws",
    "ipv4": "192.168.0.171"
  }
}
```
and the related X-Ray segment:
```
{
  "trace_id": "1-5b07b67f-ed3a0b4570cbdfe07b779e32",
  "parent_id": "02e913691d305aa0",
  "id": "f3d698d7b353651f",
  "type": "subsegment",
  "name": "unknown",
  "start_time": 1.527232127972476E9,
  "end_time": 1.527232127996071E9
}
```
This is correctly not reported as remote segment but the name is unknown. Of all I believe this is the most wrong one as it should definitely use the local service name.

## Fourth span (the real HTTP client)
```
{
  "traceId": "5b07b67fed3a0b4570cbdfe07b779e32",
  "parentId": "f3d698d7b353651f",
  "id": "164d63874559fbdf",
  "kind": "CLIENT",
  "name": "get",
  "timestamp": 1527232127973431,
  "duration": 22109,
  "localEndpoint": {
    "serviceName": "security-gateway-aws",
    "ipv4": "192.168.0.171"
  },
  "tags": {
    "http.method": "GET",
    "http.path": "/test-app/v1/public"
  }
}
```
and the X-Ray segment:
```
{
  "trace_id": "1-5b07b67f-ed3a0b4570cbdfe07b779e32",
  "parent_id": "f3d698d7b353651f",
  "id": "164d63874559fbdf",
  "type": "subsegment",
  "namespace": "remote",
  "name": "unknown",
  "start_time": 1.527232127973431E9,
  "end_time": 1.52723212799554E9,
  "http": {
    "request": {
      "method": "GET"
    }
  },
  "annotations": {
    "operation": "get",
    "http_path": "/test-app/v1/public"
  }
}
```
Same situation as for the second span.

Reply:
I'm wondering, as the AWS X-Ray documentation (https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html) states that for subsegments the name should be "The logical name of the subsegment. For downstream calls, name the subsegment after the resource or service called. For custom subsegments, name the subsegment after the code that it instruments (e.g., a function name)." should the span name be used as subsegment name for spans that don't have any kind set and use the current unknown (or the local service name) for CLIENT spans only?